cudax/include/cuda/experimental/__container/heterogeneous_iterator.cuh

```diff
@@ -0,0 +1,435 @@
+      : __ptr_(__ptr)
+  {}
+
+  //! @brief Dereference a \c heterogeneous_iterator
+  //! @return A reference to the element the iterator points to
+  _CCCL_NODISCARD _CCCL_DEVICE constexpr reference operator*() const noexcept
+  {
+    return *__ptr_;
+  }
+
+  //! @brief Operator arrow on a \c heterogeneous_iterator
+  //! @return A pointer to the element the iterator points to
+  _CCCL_NODISCARD _CCCL_DEVICE constexpr pointer operator->() const noexcept
+  {
+    return __ptr_;
+  }
+
+  //! @brief Dereference a \c heterogeneous_iterator
+  //! @param __count The offset at which we want to dereference
+  //! @return A reference of the \p __count th element after the one the iterator points to
+  _CCCL_NODISCARD _CCCL_DEVICE constexpr reference operator[](const difference_type __count) const noexcept
+  {
+    return *(__ptr_ + __count);
+  }
+
+protected:
+  pointer __ptr_ = nullptr;
+
+  template <class, bool, class...>
+  friend class heterogeneous_iterator;
+};
+
+template <class _Tp, bool _IsConst>
+class __heterogeneous_iterator_access<_Tp, _IsConst, _ExecutionSpace::__host_device>
+{
+public:
+#  if _CCCL_STD_VER >= 2014
+  using iterator_concept = _CUDA_VSTD::contiguous_iterator_tag;
+#  endif // _CCCL_STD_VER >= 2014
+  using iterator_category = _CUDA_VSTD::random_access_iterator_tag;
+  using value_type        = _Tp;
+  using difference_type   = _CUDA_VSTD::ptrdiff_t;
+  using pointer           = _CUDA_VSTD::__maybe_const<_IsConst, _Tp>*;
+  using reference         = _CUDA_VSTD::__maybe_const<_IsConst, _Tp>&;
+
+  __heterogeneous_iterator_access() = default;
+
+  _CCCL_HOST_DEVICE explicit constexpr __heterogeneous_iterator_access(pointer __ptr) noexcept
+      : __ptr_(__ptr)
+  {}
+
+  //! @brief Dereference a \c heterogeneous_iterator
+  //! @return A reference to the element the iterator points to
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr reference operator*() const noexcept
+  {
+    return *__ptr_;
+  }
+
+  //! @brief Operator arrow on a \c heterogeneous_iterator
+  //! @return A pointer to the element the iterator points to
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr pointer operator->() const noexcept
+  {
+    return __ptr_;
+  }
+
+  //! @brief Dereference a \c heterogeneous_iterator
+  //! @param __count The offset at which we want to dereference
+  //! @return A reference to the `__count` element after the one the iterator points to
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr reference operator[](const difference_type __count) const noexcept
+  {
+    return *(__ptr_ + __count);
+  }
+
+protected:
+  pointer __ptr_ = nullptr;
+
+  template <class, bool, class...>
+  friend class heterogeneous_iterator;
+};
+
+template <class _Tp, bool _IsConst, class... _Properties>
+class heterogeneous_iterator
+    : public __heterogeneous_iterator_access<_Tp, _IsConst, __select_execution_space<_Properties...>>
+{
+public:
+#  if _CCCL_STD_VER >= 2014
```

Comment:
This is because we backport the C++20 iterator machinery to C++14 and only restrict actual range CPOs + concepts to C++17

So this is actually correct